### PR TITLE
Remove repo unit index

### PR DESCRIPTION
**Pull request by Morlinest**
~Fix `Index` values in `RepoUnit`.~~#2601 has to be merged first.~

~Edit: #2601 or #2603 can fix problems with accessing repo pages with this changes.~

Removes unused `RepoUnit.Index`

### CODE REVIEW

**Comment by ethantkoenig** [comment]
@Morlinest In the future, could you please not create PRs than depend on other, unmerged PRs? If you don't create the PR until the things it depends on are finished, that would make things a lot simpler for reviewers.

**Comment by Morlinest** [comment]
@ethantkoenig It doesn't "need" anything, I want to merge another PR only because there is another bug that will pop up after merge. I know it only because I've tested it and also know, there is already fix for it. I think #2603 will also fix it. I am writing comments about merges as bug prevention...

Edit: I think it's totally legal and common to reference other issues/pullrequests.

**Comment by ethantkoenig** [comment]
@Morlinest So this PR fixes a bug that does not currently exist, but will exist after #2601 is merged?

File: models/migrations/v45.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+	units := map[UnitType]int{
+		UnitTypeCode:            0,
+		UnitTypeIssues:          1,
+		UnitTypeExternalTracker: 1,
+		UnitTypePullRequests:    2,
+		UnitTypeReleases:        3,
+		UnitTypeWiki:            4,
+		UnitTypeExternalWiki:    4,
+	}
```
Comment (daviian) [inline models/migrations/v45.go]: Why is this different than creating the RepoUnits!?
UnitTypeExternalWiki is here 4 but in when saving settings it is the Idx of models.Units, thus 6?
Definitely not correct, needs to be changed.

**Review by daviian — CHANGES_REQUESTED** [review]

**Comment by daviian** [comment]
@ethantkoenig It is a bug, that has no visible impact currently, but the `Index` of RepoUnits is indeed set to a wrong value.

**Comment by Morlinest** [comment]
@ethantkoenig Please look at files changed. Do you see 2 different ways of setting `Index` value (first time as index from range -> 0 to x, second time as int value of UnitType -> 1 to x)? Read comments in #2603. None of the current ways is correct. So this PR fixes 3 bugs at once in code + set correct values in DB.

File: models/migrations/v45.go (hunk 1 above)
Comment (Morlinest) [reply models/migrations/v45.go]: @daviian I think this is correct and doesn't need any changes. `RepoUnit` contains `UnitType`, see line below.
Please see "unit.go" file. It's copied from there. The only thing I did was to replace `map[UnitType]Unit` for `map[UnitType]int` to reduce number of lines added (around 60 lines) and copied only `Unit.Idx` values. All `Index` values in `RepoUnit` table with `UnitTypeExternalWiki` should be updated to `4` as it is in `units` map. Where do you see 6?

**Review by daviian — APPROVED** [review]

**Comment by ethantkoenig** [comment]
This may be a stupid question, but where exactly is the `RepoUnit.Index` field used? It is never read in the code (only populated in struct declarations), and it doesn't seem to be used anywhere in the HTML templates.

Also, if the index should be uniquely determined by the `RepoUnit.Type` field (as it is in this PR), do we even need to store in the database at all?

**Comment by Morlinest** [comment]
@ethantkoenig That's not a stupid question. As I wrote before in #2603

> We can improve it later. And then we can just remove Index from DB as it will be useless.

it can be removed. But first we need to check code and rewrite Home handler at least. I was focused on making quick fix of bug instead. But maybe I can rewrite #2601 to fix `UnitTypeCode` check and also to use `Unit.Idx` for rest. This should be just fast fix, but now I have time to work on it.

**Comment by Morlinest** [comment]
@ethantkoenig Changed this PR to completly remove `RepoUnit.Index` instead of fixing it as #2601 is rewriten and it's not needed for anything.

@daviian With this two changes (if aprooved and merged) will not be #2603 needed anymore. Right?

File: models/migrations/v45.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+package migrations
+
+import (
+	"fmt"
+
+	"github.com/go-xorm/xorm"
+)
+
+func removeIndexColumnFromRepoUnitTable(x *xorm.Engine) (err error) {
+	if _, err := x.Exec("ALTER TABLE repo_unit DROP COLUMN index"); err != nil {
```
Comment (lafriks) [inline models/migrations/v45.go]: Columns can not be dropped on sqlite. Please add if to switch like in this migration: https://github.com/go-gitea/gitea/blob/1739e84ac02c0384c04576a00abab9348293f9c7/models/migrations/v40.go#L43

**Review by lafriks — CHANGES_REQUESTED** [review]

File: models/migrations/v45.go (hunk 2 above)
Comment (Morlinest) [reply models/migrations/v45.go]: Done

**Review by ethantkoenig — APPROVED** [review]

**Comment by ethantkoenig** [comment]
LGTM

**Review by lafriks — APPROVED** [review]

**Comment by lafriks** [comment]
LGTM

**Comment by lunny** [comment]
conflicted

**Comment by lafriks** [comment]
@lunny fixed

**Pull request merged**